File: frontend/src/types/quizAttempt.ts
import type { QuestionType } from './quiz'

export type AttemptStatus = 'in_progress' | 'completed'

export interface QuizAttempt {
  id: string
  quiz_id: string
  quiz_title: string
  subject_id: string | null
  status: AttemptStatus
  score: number | null
  total_questions: number
  percentage_score: number | null
  started_at: string
  completed_at: string | null
}

export interface QuizAttemptQuestionResult {
  question_id: string
  order_index: number
  question_type: QuestionType
  prompt: string
  options: string[]
  submitted_answer: string | null
  is_correct: boolean | null
  correct_answer: string | null
  explanation: string | null
}

export interface QuizAttemptDetail extends QuizAttempt {
  answers: QuizAttemptQuestionResult[]
}

export interface SubmitAnswerRequest {
  submitted_answer: string
}

export interface ListAttemptsParams {
  quiz_id?: string
  subject_id?: string
}

File: backend/app/schemas/quiz_attempt.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class QuizAttemptAnswerSubmit(BaseModel):
    submitted_answer: str = Field(default="", max_length=2000)


class QuizAttemptQuestionResult(BaseModel):
    question_id: uuid.UUID
    order_index: int
    question_type: str
    prompt: str
    options: list[str]
    submitted_answer: str | None
    is_correct: bool | None
    correct_answer: str | None
    explanation: str | None


class QuizAttemptRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    quiz_id: uuid.UUID
    quiz_title: str
    subject_id: uuid.UUID | None
    status: str
    score: int | None
    total_questions: int
    percentage_score: int | None
    started_at: datetime
    completed_at: datetime | None


class QuizAttemptDetail(QuizAttemptRead):
    answers: list[QuizAttemptQuestionResult]